### ISSUE TITLE
[Train] User exceptions not propagated from remote cluster

### ISSUE THREAD

**Issue by mlappelbaum** [ray-project/ray#31913]
### What happened + What you expected to happen

Ray Trainer does not propagate exceptions from `train_loop_per_worker` when workers run on a remote cluster.  Specifically, when training on a shared cluster and submitting the training job using the ray client in a driver script running on a separate machine.  The workers will print the exception and stack trace to ray worker logs but the driver will not raise an exception, and the `Result` returned from the call to `Trainer.fit()` has `error=None`.

The expected behavior is that user exceptions are re-raised in the driver script.  This occurs when the trainer is run in a local ray cluster (i.e. when calling `ray.init()` or `ray.init('auto')`), and occurs in all cases in ray versions < 2.0.0.

### Versions / Dependencies

Ray: >= 2.0.0 (tested with 2.0.0, 2.1.0, and 2.2.0) 
Python: 3.7.11
OS: Ubuntu 18.04.6 LTS (Bionic Beaver)

### Reproduction script

1. On one host run `ray start --head --port 6379`.

2. On a different host set environment variable `RAY_CLUSTER_IP` to the ip address of the first host and run:

```python
import ray
from ray.train.data_parallel_trainer import DataParallelTrainer
from ray.air.config import ScalingConfig, RunConfig, FailureConfig
import os

# ip address of remote cluster head node
cluster_ip = os.getenv("RAY_CLUSTER_IP")

def train_func():
  raise RuntimeError("User error")

ray.init(address=f"ray://{cluster_ip}:10001")

trainer = DataParallelTrainer(
  train_loop_per_worker=train_func,
  scaling_config=ScalingConfig(num_workers=1),
  run_config=RunConfig(
    # Setting fail_fast=True does not work.
    # Setting fail_fast='raise' works but is discouraged 
    failure_config=FailureConfig(fail_fast=True),
  ),
)

try:
  result = trainer.fit()
  print("No exception raised")
  if result.error is None:
    print("Result indicates no error")
  else:
    print(f"Result contains error {result.error}")
except Exception as e:
  print(f"Trainer.fit() raised exception: {e}")
```

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by cszhu** [comment]
This P2 issue has seen no activity in the past 2 years. It will be closed in 2 weeks as part of ongoing cleanup efforts.

Please comment and remove the `pending-cleanup` label if you believe this issue should remain open.

Thanks for contributing to Ray!
